README
NAME

    Text::SpeedyFx - tokenize/hash large amount of strings efficiently

VERSION

    version 0.011

SYNOPSIS

        use Data::Dumper;
        use Text::SpeedyFx;
    
        my $sfx = Text::SpeedyFx->new;
    
        my $words_bag = $sfx->hash('To be or not to be?');
        print Dumper $words_bag;
        #$VAR1 = {
        #          '1422534433' => '1',
        #          '4120516737' => '2',
        #          '1439817409' => '2',
        #          '3087870273' => '1'
        #        };
    
        my $feature_vector = $sfx->hash_fv("thats the question", 8);
        print unpack('b*', $feature_vector);
        # 01001000

DESCRIPTION

    XS implementation of a very fast combined parser/hasher which works
    well on a variety of bag-of-word problems.

    Original implementation
    <http://www.hpl.hp.com/techreports/2008/HPL-2008-91R1.pdf> is in Java
    and was adapted for a better Unicode compliance.

METHODS

 new([$seed, $bits])

    Initialize parser/hasher, can be customized with the options:

    $seed

      Hash seed (default: 1).

    $bits

      How many bits do represent one character. The default value, 8,
      sacrifices Unicode handling but is fast and low on memory footprint.
      The value of 18 encompasses Basic Multilingual, Supplementary
      Multilingual and Supplementary Ideographic planes. See also "UNICODE
      SUPPORT"

 hash($octets)

    Parses $octets and returns a hash reference (not exactly; see "CAVEAT")
    where keys are the hashed tokens and values are their respective count.
    $octets are assumed to represent UTF-8 string unless Text::SpeedyFx is
    instantiated with "$bits" == 8 (which forces Latin-1 mode, see "UNICODE
    SUPPORT"). Note that this is the slowest form due to the
    (computational) complexity of the associative array
    <https://en.wikipedia.org/wiki/Associative_array> data structure
    itself: hash_fv()/hash_min() variants are up to 260% faster!

 hash_fv($octets, $n)

    Parses $octets and returns a feature vector (string of bits) with
    length $n. $n is supposed to be a multiplier of 8, as the length of the
    resulting feature vector is ceil($n / 8). See the included utilities
    cosine_cmp and uniq_wc.

 hash_min($octets)

    Parses $octets and returns the hash with the lowest value. Useful in
    MinHash <http://en.wikipedia.org/wiki/MinHash> implementation. See also
    the included minhash_cmp utility.

UNICODE SUPPORT

    Due to the nature of Perl, Unicode support is handled differently from
    the original implementation. By default, Text::SpeedyFx recognizes
    UTF-8 encoded code points in the range 00000-2FFFF:

      * Plane 0, the Basic Multilingual Plane (BMP, 0000–FFFF)

      * Plane 1, the Supplementary Multilingual Plane (SMP, 10000–1FFFF)

      * Plane 2, the Supplementary Ideographic Plane (SIP, 20000–2FFFF)

      * There are planes up to 16; however, as in Perl v5.16.2, there are
      no code points matching isALNUM_utf8() there (so it's irrelevant for
      proper algorithm operation).

    Although, there is a major drawback: in this mode, each instance
    allocates up to 1 MB of memory.

    If the application doesn't need to support code points beyond the Plane
    0 (like the original SpeedyFx implementation) it is possible to
    constraint the address space to 16 bits, which lowers memory allocation
    to up to 256 KB. In fact, Text::SpeedyFx constructor accepts bit range
    between 8 and 18 to address code points.

 LATIN-1 SUPPORT

    8 bit address space has one special meaning: it completely disables
    multibyte support. In 8 bit mode, each instance will only allocate 256
    bytes and hashing will run up to 340% faster! Tokenization will
    fallback to ISO 8859-1 West European languages (Latin-1) character
    definitions.

BENCHMARK

    The test platform configuration:

      * Intel® Xeon® E5620 CPU @ 2.40GHz (similar to the one cited in the
      reference paper);

      * Debian 6.0.6 (Squeeze, 64-bit);

      * Perl v5.16.1 (installed via perlbrew);

      * enwik8 from the Large Text Compression Benchmark
      <https://cs.fit.edu/~mmahoney/compression/text.html>.

                           Rate murmur_utf8 hash_utf8 hash_min_utf8  hash hash_fv hash_min
        murmur_utf8      6 MB/s          --      -79%          -86%  -89%    -97%     -97%
        hash_utf8       30 MB/s        376%        --          -35%  -47%    -84%     -85%
        hash_min_utf8   47 MB/s        637%       55%            --  -18%    -76%     -77%
        hash            58 MB/s        803%       90%           23%    --    -70%     -72%
        hash_fv        194 MB/s       2946%      541%          313%  237%      --      -6%
        hash_min       206 MB/s       3143%      582%          340%  259%      6%       --

    All the tests except the ones with _utf8 suffix were made in Latin-1
    mode. For comparison, murmur_utf8 was implemented using
    Digest::MurmurHash hasher and native regular expression tokenizer:

        ++$fv->{murmur_hash(lc $1)}
            while $data =~ /(\w+)/gx;

    See also the eg/benchmark.pl script.

CAVEAT

    For performance reasons, hash() method returns a tied hash which is an
    interface to nedtries
    <http://www.nedprod.com/programs/portable/nedtries/>. The interesting
    property of a trie data structure <https://en.wikipedia.org/wiki/Trie>
    is that the keys are "nearly sorted" (and the first key is guaranteed
    to be the lowest), so:

        # This:
        $fv = $sfx->hash($data);
        ($min) = each %$fv;
        # Is the same as this:
        ($min) = $sfx->hash_min($data);
        # (albeit the later being 2x faster)

    The downside is the magic involved, the delete breaking the key order,
    and the memory usage. The hardcoded limit is 524288 unique keys per
    result, which consumes ~25MB of RAM on a 64-bit architecture. Exceeding
    this will croak with the message "too many unique tokens in a single
    data chunk". The only way to raise this limit is by recompilation of
    the XS module:

        perl Makefile.PL DEFINE=-DMAX_TRIE_SIZE=2097152
        make
        make test
        make install

REFERENCES

      * Extremely Fast Text Feature Extraction for Classification and
      Indexing <http://www.hpl.hp.com/techreports/2008/HPL-2008-91R1.pdf>
      by George Forman <http://www.hpl.hp.com/personal/George_Forman/> and
      Evan Kirshenbaum <http://www.kirshenbaum.net/evan/index.htm>

      * MinHash — выявляем похожие множества
      <http://habrahabr.ru/post/115147/>

      * Фильтр Блума <http://habrahabr.ru/post/112069/>

      * cosine_cmp, minhash_cmp and uniq_wc utilities from this
      distribution

AUTHOR

    Stanislaw Pusep <stas@sysd.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Stanislaw Pusep.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Text-SpeedyFx

0.011 2014-06-20T12:57:24
 - Dist::Zilla maintenance (Stanislaw Pusep)
 - Documentation update (Stanislaw Pusep)
 - drop PWP::Encoding dependency (Sergey Romanov)

0.010 2013-01-09T14:33:25
 - documentation improvement (Stanislaw Pusep)
 - fixed undefined symbol: DPPP_my_newSVpvn_flags (Stanislaw Pusep)
 - nedtrie CLEAR test (Stanislaw Pusep)

0.009 2013-01-06T22:37:32
 - nedtrie test (Stanislaw Pusep)
 - speedup (Stanislaw Pusep)
 - documented the nedtries caveat (Stanislaw Pusep)
 - safety check (Stanislaw Pusep)
 - DELETE implemented (Stanislaw Pusep)
 - README.pod update (Stanislaw Pusep)
 - benchmark update (Stanislaw Pusep)
 - major reorganization (Stanislaw Pusep)
 - ++magic (Stanislaw Pusep)
 - tied implementation complete (Stanislaw Pusep)
 - tied NEDTRIE interface (Stanislaw Pusep)
 - removed unnecessary speedyfx_store() (Stanislaw Pusep)
 - realloc() breaks pointers on Linux (Stanislaw Pusep)
 - ++overengineering (Stanislaw Pusep)

0.008 2013-01-01T21:36:52
 - renamed cosine_sim to cosine_cmp (to keep consistence with minhash_cmp)
   (Stanislaw Pusep)
 - fixed function name clash (_itoa already present on Win32) (Stanislaw
   Pusep)

0.007 2012-12-31T01:50:24
 - updated benchmark results (yep, it is that faster, even on the lower
   clock! (Stanislaw Pusep)
 - less indirections (Stanislaw Pusep)
 - micro-optimizations for latin1 (Stanislaw Pusep)
 - macro rampage (Stanislaw Pusep)
 - macros cleanup (Stanislaw Pusep)
 - .gitignore update (Stanislaw Pusep)
 - no more warnings under Clang (Stanislaw Pusep)
 - added murmur to benchmark for comparison (Stanislaw Pusep)
 - replaced sprintf() by custom itoa() (Stanislaw Pusep)

0.006 2012-12-27T22:46:31
 - minor fixes (Stanislaw Pusep)
 - documented multibyte parsing tricks (Stanislaw Pusep)
 - break test (Stanislaw Pusep)
 - leak test (Stanislaw Pusep)
 - more optimizations (Stanislaw Pusep)
 - better utf8/latin1 tests (Stanislaw Pusep)
 - updated benchmark results (Stanislaw Pusep)
 - added cosine_sim utility (Stanislaw Pusep)
 - reallocation fixed (Stanislaw Pusep)
 - graduated the uniq_wc tool (Stanislaw Pusep)
 - make use of PerlIO::mmap layer (Stanislaw Pusep)
 - examples cleanup (Stanislaw Pusep)
 - optimizations (Stanislaw Pusep)
 - implemented variable codetable (+ raw ASCII support) (Stanislaw Pusep)
 - variable size arrays (Stanislaw Pusep)
 - Dist::Zilla profile update (Stanislaw Pusep)
 - File::Slurp => File::Map (Stanislaw Pusep)

0.005 2012-04-10T13:58:26
 - feature vector reimplemented as a bit string (Stanislaw Pusep)
 - Changelog (Stanislaw Pusep)

0.004 2012-03-30T14:55:01
 - minhash_cmp @ARGV fix (Stanislaw Pusep)

0.003 2012-03-30T14:39:20
 - oops! (Stanislaw Pusep)
 - cast hash keys as unsigned int (Stanislaw Pusep)
 - test separation (Stanislaw Pusep)
 - minhash_cmp documentation (Stanislaw Pusep)
 - eg/benchmark.pl (Stanislaw Pusep)

0.002 2012-03-29T13:24:38
 - perl v5.10.2 compatibility (Stanislaw Pusep)
 - static fold table (Stanislaw Pusep)
 - hash_min() (Stanislaw Pusep)
 - min method (Stanislaw Pusep)
 - ++MinHash utility (Stanislaw Pusep)

0.001 2012-03-28T22:12:52
 - ++documentation (Stanislaw Pusep)
 - cleanup (Stanislaw Pusep)
 - seed is optional (Stanislaw Pusep)
 - ++hash_fv (Stanislaw Pusep)
 - passes release tests (Stanislaw Pusep)
 - INIT (Stanislaw Pusep)



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "tokenize/hash large amount of strings efficiently",
  "AUTHOR" => "Stanislaw Pusep <stas\@sysd.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Text-SpeedyFx",
  "EXE_FILES" => [
    "bin/cosine_cmp",
    "bin/minhash_cmp",
    "bin/uniq_wc"
  ],
  "LICENSE" => "perl",
  "NAME" => "Text::SpeedyFx",
  "PREREQ_PM" => {
    "List::MoreUtils" => 0,
    "autodie" => 0
  },
  "VERSION" => "0.011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "List::MoreUtils" => 0,
  "autodie" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Text/SpeedyFx.pm
package Text::SpeedyFx;
# ABSTRACT: tokenize/hash large amount of strings efficiently

use strict;
use utf8;
use warnings;

use base q(Exporter);

our $VERSION = '0.011'; # VERSION

require XSLoader;
XSLoader::load('Text::SpeedyFx', $VERSION);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Text::SpeedyFx - tokenize/hash large amount of strings efficiently

=head1 VERSION

version 0.011

=head1 SYNOPSIS

    use Data::Dumper;
    use Text::SpeedyFx;

    my $sfx = Text::SpeedyFx->new;

    my $words_bag = $sfx->hash('To be or not to be?');
    print Dumper $words_bag;
    #$VAR1 = {
    #          '1422534433' => '1',
    #          '4120516737' => '2',
    #          '1439817409' => '2',
    #          '3087870273' => '1'
    #        };

    my $feature_vector = $sfx->hash_fv("thats the question", 8);
    print unpack('b*', $feature_vector);
    # 01001000

=head1 DESCRIPTION

XS implementation of a very fast combined parser/hasher which works well on a variety of I<bag-of-word> problems.

L<Original implementation|http://www.hpl.hp.com/techreports/2008/HPL-2008-91R1.pdf> is in Java and was adapted for a better Unicode compliance.

=head1 METHODS

=head2 new([$seed, $bits])

Initialize parser/hasher, can be customized with the options:

=over 4

=item C<$seed>

Hash seed (default: 1).

=item C<$bits>

How many bits do represent one character.
The default value, 8, sacrifices Unicode handling but is fast and low on memory footprint.
The value of 18 encompasses I<Basic Multilingual>, I<Supplementary Multilingual> and I<Supplementary Ideographic> planes.
See also L</UNICODE SUPPORT>

=back

=head2 hash($octets)

Parses C<$octets> and returns a hash reference (not exactly; see L</CAVEAT>) where keys are the hashed tokens and values are their respective count.
C<$octets> are assumed to represent UTF-8 string unless L<Text::SpeedyFx> is instantiated with L</$bits> == 8
(which forces Latin-1 mode, see L</UNICODE SUPPORT>).
Note that this is the slowest form due to the (computational) complexity of the L<associative array|https://en.wikipedia.org/wiki/Associative_array> data structure itself:
C<hash_fv()>/C<hash_min()> variants are up to 260% faster!

=head2 hash_fv($octets, $n)

Parses C<$octets> and returns a feature vector (string of bits) with length C<$n>.
C<$n> is supposed to be a multiplier of 8, as the length of the resulting feature vector is C<ceil($n / 8)>.
See the included utilities L<cosine_cmp> and L<uniq_wc>.

=head2 hash_min($octets)

Parses C<$octets> and returns the hash with the lowest value.
Useful in L<MinHash|http://en.wikipedia.org/wiki/MinHash> implementation.
See also the included L<minhash_cmp> utility.

=head1 UNICODE SUPPORT

Due to the nature of Perl, Unicode support is handled differently from the original implementation.
By default, L<Text::SpeedyFx> recognizes UTF-8 encoded code points in the range I<00000-2FFFF>:

=over 4

=item *

B<Plane 0>, the B<Basic Multilingual Plane> (BMP, I<0000–FFFF>)

=item *

B<Plane 1>, the B<Supplementary Multilingual Plane> (SMP, I<10000–1FFFF>)

=item *

B<Plane 2>, the B<Supplementary Ideographic Plane> (SIP, I<20000–2FFFF>)

=item *

There are planes up to 16; however, as in Perl v5.16.2, there are no code points matching C<isALNUM_utf8()> there (so it's irrelevant for proper algorithm operation).

=back

Although, there is a major drawback: in this mode, B<each instance> allocates up to 1 MB of memory.

If the application doesn't need to support code points beyond the B<Plane 0>
(like the original SpeedyFx implementation) it is possible to constraint the address space to 16 bits, which lowers memory allocation to up to 256 KB.
In fact, L<Text::SpeedyFx> constructor accepts bit range between 8 and 18 to address code points.

=head2 LATIN-1 SUPPORT

8 bit address space has one special meaning: it completely disables multibyte support.
In 8 bit mode, each instance will only allocate 256 bytes and hashing will run up to 340% faster!
Tokenization will fallback to I<ISO 8859-1 West European languages (Latin-1)> character definitions.

=head1 BENCHMARK

The test platform configuration:

=over 4

=item *

Intel® Xeon® E5620 CPU @ 2.40GHz (similar to the one cited in the reference paper);

=item *

Debian 6.0.6 (Squeeze, 64-bit);

=item *

Perl v5.16.1 (installed via L<perlbrew>);

=item *

F<enwik8> from the L<Large Text Compression Benchmark|https://cs.fit.edu/~mmahoney/compression/text.html>.

=back

                       Rate murmur_utf8 hash_utf8 hash_min_utf8  hash hash_fv hash_min
    murmur_utf8      6 MB/s          --      -79%          -86%  -89%    -97%     -97%
    hash_utf8       30 MB/s        376%        --          -35%  -47%    -84%     -85%
    hash_min_utf8   47 MB/s        637%       55%            --  -18%    -76%     -77%
    hash            58 MB/s        803%       90%           23%    --    -70%     -72%
    hash_fv        194 MB/s       2946%      541%          313%  237%      --      -6%
    hash_min       206 MB/s       3143%      582%          340%  259%      6%       --

All the tests except the ones with C<_utf8> suffix were made in L<Latin-1 mode|/UNICODE SUPPORT>.
For comparison, C<murmur_utf8> was implemented using L<Digest::MurmurHash> hasher and native regular expression tokenizer:

    ++$fv->{murmur_hash(lc $1)}
        while $data =~ /(\w+)/gx;

See also the F<eg/benchmark.pl> script.

=head1 CAVEAT

For performance reasons, C<hash()> method returns a L<tied hash|perltie> which is an interface to
L<nedtries|http://www.nedprod.com/programs/portable/nedtries/>.
The interesting property of a L<trie data structure|https://en.wikipedia.org/wiki/Trie>
is that the keys are "nearly sorted" (and the first key is guaranteed to be the lowest), so:

    # This:
    $fv = $sfx->hash($data);
    ($min) = each %$fv;
    # Is the same as this:
    ($min) = $sfx->hash_min($data);
    # (albeit the later being 2x faster)

The downside is the magic involved, the C<delete> breaking the key order, and the memory usage.
The hardcoded limit is 524288 unique keys per result, which consumes ~25MB of RAM on a 64-bit architecture.
Exceeding this will C<croak> with the message I<"too many unique tokens in a single data chunk">.
The only way to raise this limit is by recompilation of the XS module:

    perl Makefile.PL DEFINE=-DMAX_TRIE_SIZE=2097152
    make
    make test
    make install

=head1 REFERENCES

=over 4

=item *

L<Extremely Fast Text Feature Extraction for Classification and Indexing|http://www.hpl.hp.com/techreports/2008/HPL-2008-91R1.pdf> by L<George Forman|http://www.hpl.hp.com/personal/George_Forman/> and L<Evan Kirshenbaum|http://www.kirshenbaum.net/evan/index.htm>

=item *

L<MinHash — выявляем похожие множества|http://habrahabr.ru/post/115147/>

=item *

L<Фильтр Блума|http://habrahabr.ru/post/112069/>

=item *

L<cosine_cmp>, L<minhash_cmp> and L<uniq_wc> utilities from this distribution

=back

=head1 AUTHOR

Stanislaw Pusep <stas@sysd.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Stanislaw Pusep.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
